### ISSUE TITLE
Mixlib not compatible with Chef 12.4.3? 

### ISSUE THREAD

**Issue by mckownam** [chef-cookbooks/chef_client_updater#9]
### Cookbook version
1.0.2

### Chef-client version
Multiple; 12.4+ (up to latest: 13.0.118)

### Platform Details
Ubuntu 12.04 x64; Running on VMware. 

### Scenario:
Force a downgrade from Chef 12.16.42 to 12.4.3 and then upgrade to any higher version to test upgrade and downgrade capabilities. Downgrade works just fine, but upgrade exits with Mixlib::ShellOut::ShellCommandFailed error. 

### Steps to Reproduce:
Take a Ubuntu 12.04 x64 VM on VMware and run the chef_client_updater cookbook to downgrade it to chef 12.4.3. Then attempt to upgrade to a newer version (like 12.5.1 or 12.16.42 or 13.0.118). The same Mixlib::ShellOut::ShellCommandFailed error appears in all of those. 

### Expected Result:
Since the cookbook claims compatibility with Chef 12.1+, the cookbook should successfully kill the chef run and upgrade the instance so that the next chef run is the upgraded version. 

### Actual Result:
Output from the Chef run:
<pre>
Recipe: chef_client_updater::default
  * chef_client_updater_updater[update chef-client] action update
    * chef_gem[mixlib-install] action install

      ================================================================================
      Error executing action `install` on resource 'chef_gem[mixlib-install]'
      ================================================================================

      Mixlib::ShellOut::ShellCommandFailed
      ------------------------------------
      Expected process to exit with [0], but received '1'
      ---- Begin output of /opt/chef/embedded/bin/gem install mixlib-install -q --no-rdoc --no-ri -v "3.2.0" ----
      STDOUT:
      STDERR: ERROR:  While executing gem ... (TypeError)
          no implicit conversion of nil into String
      ---- End output of /opt/chef/embedded/bin/gem install mixlib-install -q --no-rdoc --no-ri -v "3.2.0" ----
      Ran /opt/chef/embedded/bin/gem install mixlib-install -q --no-rdoc --no-ri -v "3.2.0" returned 1

      Cookbook Trace:
      ---------------
      /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/resource_builder.rb:106:in `build'
      /var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:45:in `rescue in load_mixlib_install'
      /var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:42:in `load_mixlib_install'
      /var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:64:in `mixlib_install'
      /var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:33:in `block (2 levels) in class_from_file'
      /var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:32:in `block in class_from_file'
      /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:132:in `instance_eval'
      /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:132:in `compile_and_converge_action'
      /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:78:in `run_action'
      /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `block (2 levels) in converge'
      /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `each'
      /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `block in converge'
      /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:105:in `converge'

      Resource Declaration:
      ---------------------
      # In /var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb

       45:     chef_gem 'mixlib-install' do
       46:       compile_time true if respond_to?(:compile_time)
       47:     end
       48:

      Compiled Resource:
      ------------------
      # Declared in /var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:45:in `rescue in load_mixlib_install'

      chef_gem("mixlib-install") do
        action [:install]
        retries 0
        retry_delay 2
        default_guard_interpreter :default
        package_name "mixlib-install"
        version "3.2.0"
        compile_time true
        gem_binary "/opt/chef/embedded/bin/gem"
        declared_type :chef_gem
        cookbook_name "chef_client_updater"
      end

    ================================================================================
    Error executing action `update` on resource 'chef_client_updater_updater[update chef-client]'
    ================================================================================

    Mixlib::ShellOut::ShellCommandFailed
    ------------------------------------
    chef_gem[mixlib-install] (/var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb line 45) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
    ---- Begin output of /opt/chef/embedded/bin/gem install mixlib-install -q --no-rdoc --no-ri -v "3.2.0" ----
    STDOUT:
    STDERR: ERROR:  While executing gem ... (TypeError)
        no implicit conversion of nil into String
    ---- End output of /opt/chef/embedded/bin/gem install mixlib-install -q --no-rdoc --no-ri -v "3.2.0" ----
    Ran /opt/chef/embedded/bin/gem install mixlib-install -q --no-rdoc --no-ri -v "3.2.0" returned 1

    Cookbook Trace:
    ---------------
    /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/resource_builder.rb:106:in `build'
    /var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:45:in `rescue in load_mixlib_install'
    /var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:42:in `load_mixlib_install'
    /var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:64:in `mixlib_install'
    /var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:33:in `block (2 levels) in class_from_file'
    /var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:32:in `block in class_from_file'
    /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:132:in `instance_eval'
    /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:132:in `compile_and_converge_action'
    /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:78:in `run_action'
    /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `block (2 levels) in converge'
    /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `each'
    /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `block in converge'
    /var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:105:in `converge'

    Resource Declaration:
    ---------------------
    # In /var/chef/cache/cookbooks/chef_client_updater/recipes/default.rb

     20: chef_client_updater 'update chef-client' do
     21:   channel node['chef_client_updater']['channel']
     22:   version node['chef_client_updater']['version']
     23:   prevent_downgrade node['chef_client_updater']['prevent_downgrade']
     24:   post_install_action node['chef_client_updater']['post_install_action']
     25: end

    Compiled Resource:
    ------------------
    # Declared in /var/chef/cache/cookbooks/chef_client_updater/recipes/default.rb:20:in `from_file'

    chef_client_updater_updater("update chef-client") do
      action [:update]
      retries 0
      retry_delay 2
      default_guard_interpreter :default
      declared_type :chef_client_updater
      cookbook_name "chef_client_updater"
      recipe_name "default"
      channel "stable"
      version "12.5.1"
      post_install_action "kill"
    end

Running handlers:
[2017-04-26T08:49:02-04:00] ERROR: Running exception handlers
Running handlers complete
[2017-04-26T08:49:02-04:00] ERROR: Exception handlers complete
Chef Client failed. 1 resources updated in 3.68236509 seconds
[2017-04-26T08:49:03-04:00] FATAL: Stacktrace dumped to /var/chef/cache/chef-stacktrace.out
[2017-04-26T08:49:03-04:00] ERROR: chef_client_updater_updater[update chef-client] (chef_client_updater::default line 20) had an error: Mixlib::ShellOut::ShellCommandFailed: chef_gem[mixlib-install] (/var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb line 45) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
---- Begin output of /opt/chef/embedded/bin/gem install mixlib-install -q --no-rdoc --no-ri -v "3.2.0" ----
STDOUT:
STDERR: ERROR:  While executing gem ... (TypeError)
    no implicit conversion of nil into String
---- End output of /opt/chef/embedded/bin/gem install mixlib-install -q --no-rdoc --no-ri -v "3.2.0" ----
Ran /opt/chef/embedded/bin/gem install mixlib-install -q --no-rdoc --no-ri -v "3.2.0" returned 1
[2017-04-26T08:49:03-04:00] FATAL: Chef::Exceptions::ChildConvergeError: Chef run process exited unsuccessfully (exit code 1)</pre>

**Comment by iennae** [comment]
Unable to replicate the error you are reporting. I'm replicating on ubuntu 14.04 as ubuntu 12.04 is EOL. 

<details>
-----> Starting Kitchen (v1.16.0)
-----> Creating <default-ubuntu-1404>...
       Bringing machine 'default' up with 'virtualbox' provider...
       ==> default: Box 'bento/ubuntu-14.04' could not be found. Attempting to find and install...
           default: Box Provider: virtualbox
           default: Box Version: >= 0
       ==> default: Loading metadata for box 'bento/ubuntu-14.04'
           default: URL: https://atlas.hashicorp.com/bento/ubuntu-14.04
       ==> default: Adding box 'bento/ubuntu-14.04' (v2.3.5) for provider: virtualbox
           default: Downloading: https://atlas.hashicorp.com/bento/boxes/ubuntu-14.04/versions/2.3.5/providers/virtualbox.box
==> default: Successfully added box 'bento/ubuntu-14.04' (v2.3.5) for 'virtualbox'!
       ==> default: Importing base box 'bento/ubuntu-14.04'...
==> default: Matching MAC address for NAT networking...
       ==> default: Checking if box 'bento/ubuntu-14.04' is up to date...
       ==> default: Setting the name of the VM: kitchen-chef_client_updater-default-ubuntu-1404_default_1493749795237_4035
       ==> default: Fixed port collision for 22 => 2222. Now on port 2201.
       ==> default: Clearing any previously set network interfaces...
       ==> default: Preparing network interfaces based on configuration...
           default: Adapter 1: nat
       ==> default: Forwarding ports...
           default: 22 (guest) => 2201 (host) (adapter 1)
       ==> default: Booting VM...
       ==> default: Waiting for machine to boot. This may take a few minutes...
           default: SSH address: 127.0.0.1:2201
           default: SSH username: vagrant
           default: SSH auth method: private key
           default:
           default: Vagrant insecure key detected. Vagrant will automatically replace
           default: this with a newly generated keypair for better security.
           default:
           default: Inserting generated public key within guest...
           default: Removing insecure key from the guest if it's present...
           default: Key inserted! Disconnecting and reconnecting using new SSH key...
       ==> default: Machine booted and ready!
       ==> default: Checking for guest additions in VM...
       ==> default: Setting hostname...
       ==> default: Mounting shared folders...
           default: /tmp/omnibus/cache => /Users/sigje/.kitchen/cache
       ==> default: Machine not provisioned because `--no-provision` is specified.
       [SSH] Established
       Vagrant instance <default-ubuntu-1404> created.
       Finished creating <default-ubuntu-1404> (3m4.04s).
-----> Converging <default-ubuntu-1404>...
       Preparing files for transfer
       Preparing dna.json
       Resolving cookbook dependencies with Berkshelf 5.6.4...
       Removing non-cookbook files before transfer
       Preparing validation.pem
       Preparing client.rb
-----> Installing Chef Omnibus (12.4.3)
       Downloading https://omnitruck.chef.io/install.sh to file /tmp/install.sh
       Trying wget...
       Download complete.
       ubuntu 14.04 x86_64
       Getting information for chef stable 12.4.3 for ubuntu...
       downloading https://omnitruck.chef.io/stable/chef/metadata?v=12.4.3&p=ubuntu&pv=14.04&m=x86_64
         to file /tmp/install.sh.1520/metadata.txt
       trying wget...
       sha1	f4e5203c1c0964e026eed7dc9d23a193a30a5c4e
       sha256	de772b659e09b0ead5a116585f0f610ab74c82cb313a7bf7c747a6eb94db59df
       url	https://packages.chef.io/files/stable/chef/12.4.3/ubuntu/14.04/chef_12.4.3-1_amd64.deb
       version	12.4.3
       downloaded metadata file looks valid...
       downloading https://packages.chef.io/files/stable/chef/12.4.3/ubuntu/14.04/chef_12.4.3-1_amd64.deb
         to file /tmp/omnibus/cache/chef_12.4.3-1_amd64.deb
       trying wget...
       Comparing checksum with sha256sum...
       Installing chef 12.4.3
       installing with dpkg...
       Selecting previously unselected package chef.
(Reading database ... 35394 files and directories currently installed.)
       Preparing to unpack .../cache/chef_12.4.3-1_amd64.deb ...
       Unpacking chef (12.4.3-1) ...
       Setting up chef (12.4.3-1) ...
       Thank you for installing Chef!
       Transferring files to <default-ubuntu-1404>
       Starting Chef Client, version 12.4.3
       Creating a new client identity for default-ubuntu-1404 using the validator key.
       [2017-05-02T18:30:56+00:00] WARN: Child with name 'dna.json' found in multiple directories: /tmp/kitchen/dna.json and /tmp/kitchen/dna.json
       [2017-05-02T18:30:56+00:00] WARN: Child with name 'dna.json' found in multiple directories: /tmp/kitchen/dna.json and /tmp/kitchen/dna.json
       resolving cookbooks for run list: ["chef_client_updater::default"]
       [2017-05-02T18:30:56+00:00] WARN: Child with name 'dna.json' found in multiple directories: /tmp/kitchen/dna.json and /tmp/kitchen/dna.json
       Synchronizing Cookbooks:
         - chef_client_updater
         - compat_resource
       Compiling Cookbooks...
       Converging 1 resources
       Recipe: chef_client_updater::default
         * chef_client_updater_updater[update chef-client] action update
           * chef_gem[mixlib-versioning] action install
             - install version 1.1.0 of package mixlib-versioning
           * chef_gem[mixlib-install] action install
             - install version 3.2.0 of package mixlib-install
       [2017-05-02T18:31:31+00:00] WARN: New chef-client installed. Forcing chef exit!

           ================================================================================
           Error executing action `update` on resource 'chef_client_updater_updater[update chef-client]'
           ================================================================================

           SystemExit
           ----------
           exit

           Cookbook Trace:
           ---------------
           /tmp/kitchen/cache/cookbooks/chef_client_updater/resources/updater.rb:131:in `exit'
           /tmp/kitchen/cache/cookbooks/chef_client_updater/resources/updater.rb:131:in `run_post_install_action'
           /tmp/kitchen/cache/cookbooks/chef_client_updater/resources/updater.rb:35:in `block (2 levels) in class_from_file'
           /tmp/kitchen/cache/cookbooks/chef_client_updater/resources/updater.rb:32:in `block in class_from_file'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:132:in `instance_eval'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:132:in `compile_and_converge_action'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:78:in `run_action'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `block (2 levels) in converge'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `each'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `block in converge'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:105:in `converge'

           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/chef_client_updater/recipes/default.rb

            20: chef_client_updater 'update chef-client' do
            21:   channel node['chef_client_updater']['channel']
            22:   version node['chef_client_updater']['version']
            23:   prevent_downgrade node['chef_client_updater']['prevent_downgrade']
            24:   post_install_action node['chef_client_updater']['post_install_action']
            25: end

           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/chef_client_updater/recipes/default.rb:20:in `from_file'

           chef_client_updater_updater("update chef-client") do
             action [:update]
             updated true
             updated_by_last_action true
             retries 0
             retry_delay 2
             default_guard_interpreter :default
             declared_type :chef_client_updater
             cookbook_name "chef_client_updater"
             recipe_name "default"
             channel "current"
             version "latest"
             prevent_downgrade true
             post_install_action "kill"
           end


       Running handlers:
       [2017-05-02T18:31:31+00:00] ERROR: Running exception handlers
       Running handlers complete
       [2017-05-02T18:31:31+00:00] ERROR: Exception handlers complete
       Chef Client failed. 2 resources updated in 36.592918915 seconds
       [2017-05-02T18:31:31+00:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
       Starting Chef Client, version 13.1.6
       resolving cookbooks for run list: ["chef_client_updater::default"]
       Synchronizing Cookbooks:
         - chef_client_updater (1.0.2)
         - compat_resource (12.19.0)
       Installing Cookbook Gems:
       Compiling Cookbooks...
       Converging 1 resources
       Recipe: chef_client_updater::default
         * chef_client_updater_updater[update chef-client] action update
           * chef_gem[mixlib-versioning] action install
             - install version 1.1.0 of package mixlib-versioning
           * chef_gem[mixlib-install] action install
             - install version 3.2.0 of package mixlib-install
           * chef_gem[mixlib-versioning] action install (up to date)
           * chef_gem[mixlib-install] action install (up to date)


       Running handlers:
       Running handlers complete
       Chef Client finished, 3/5 resources updated in 12 seconds
       Finished converging <default-ubuntu-1404> (1m25.51s).
</details>

Second converge and does a downgrade and upgrade:

<details>
ennifers-MBP:chef_client_updater sigje$ kitchen converge default-ubuntu-1404
-----> Starting Kitchen (v1.16.0)
-----> Converging <default-ubuntu-1404>...
       Preparing files for transfer
       Preparing dna.json
       Resolving cookbook dependencies with Berkshelf 5.6.4...
       Removing non-cookbook files before transfer
       Preparing validation.pem
       Preparing client.rb
-----> Installing Chef Omnibus (12.4.3)
       Downloading https://omnitruck.chef.io/install.sh to file /tmp/install.sh
       Trying wget...
       Download complete.
       ubuntu 14.04 x86_64
       Getting information for chef stable 12.4.3 for ubuntu...
       downloading https://omnitruck.chef.io/stable/chef/metadata?v=12.4.3&p=ubuntu&pv=14.04&m=x86_64
         to file /tmp/install.sh.2255/metadata.txt
       trying wget...
       sha1	f4e5203c1c0964e026eed7dc9d23a193a30a5c4e
       sha256	de772b659e09b0ead5a116585f0f610ab74c82cb313a7bf7c747a6eb94db59df
       url	https://packages.chef.io/files/stable/chef/12.4.3/ubuntu/14.04/chef_12.4.3-1_amd64.deb
       version	12.4.3
       downloaded metadata file looks valid...
       /tmp/omnibus/cache/chef_12.4.3-1_amd64.deb already exists, verifiying checksum...
       Comparing checksum with sha256sum...
       checksum compare succeeded, using existing file!
       Installing chef 12.4.3
       installing with dpkg...
       dpkg: warning: downgrading chef from 13.1.6-1 to 12.4.3-1
(Reading database ... 48424 files and directories currently installed.)
       Preparing to unpack .../cache/chef_12.4.3-1_amd64.deb ...
       Unpacking chef (12.4.3-1) over (13.1.6-1) ...
       dpkg: warning: unable to delete old directory '/opt/chef/embedded/lib/ruby/gems/2.4.0/gems': Directory not empty
       dpkg: warning: unable to delete old directory '/opt/chef/embedded/lib/ruby/gems/2.4.0/specifications': Directory not empty
       dpkg: warning: unable to delete old directory '/opt/chef/embedded/lib/ruby/gems/2.4.0': Directory not empty
       Setting up chef (12.4.3-1) ...
       Thank you for installing Chef!
       Transferring files to <default-ubuntu-1404>
       Starting Chef Client, version 12.4.3
       [2017-05-02T18:48:20+00:00] WARN: Child with name 'dna.json' found in multiple directories: /tmp/kitchen/dna.json and /tmp/kitchen/dna.json
       resolving cookbooks for run list: ["chef_client_updater::default"]
       Synchronizing Cookbooks:
         - chef_client_updater
         - compat_resource
       Compiling Cookbooks...
       Converging 1 resources
       Recipe: chef_client_updater::default
         * chef_client_updater_updater[update chef-client] action update[2017-05-02T18:48:57+00:00] WARN: New chef-client installed. Forcing chef exit!


           ================================================================================
           Error executing action `update` on resource 'chef_client_updater_updater[update chef-client]'
           ================================================================================

           SystemExit
           ----------
           exit

           Cookbook Trace:
           ---------------
           /tmp/kitchen/cache/cookbooks/chef_client_updater/resources/updater.rb:131:in `exit'
           /tmp/kitchen/cache/cookbooks/chef_client_updater/resources/updater.rb:131:in `run_post_install_action'
           /tmp/kitchen/cache/cookbooks/chef_client_updater/resources/updater.rb:35:in `block (2 levels) in class_from_file'
           /tmp/kitchen/cache/cookbooks/chef_client_updater/resources/updater.rb:32:in `block in class_from_file'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:132:in `instance_eval'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:132:in `compile_and_converge_action'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:78:in `run_action'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `block (2 levels) in converge'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `each'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `block in converge'
           /tmp/kitchen/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:105:in `converge'

           Resource Declaration:
           ---------------------
           # In /tmp/kitchen/cache/cookbooks/chef_client_updater/recipes/default.rb

            20: chef_client_updater 'update chef-client' do
            21:   channel node['chef_client_updater']['channel']
            22:   version node['chef_client_updater']['version']
            23:   prevent_downgrade node['chef_client_updater']['prevent_downgrade']
            24:   post_install_action node['chef_client_updater']['post_install_action']
            25: end

           Compiled Resource:
           ------------------
           # Declared in /tmp/kitchen/cache/cookbooks/chef_client_updater/recipes/default.rb:20:in `from_file'

           chef_client_updater_updater("update chef-client") do
             action [:update]
             retries 0
             retry_delay 2
             default_guard_interpreter :default
             declared_type :chef_client_updater
             cookbook_name "chef_client_updater"
             recipe_name "default"
             channel "current"
             version "latest"
             prevent_downgrade true
             post_install_action "kill"
           end


       Running handlers:
       [2017-05-02T18:48:57+00:00] ERROR: Running exception handlers
       Running handlers complete
       [2017-05-02T18:48:57+00:00] ERROR: Exception handlers complete
       Chef Client failed. 0 resources updated in 44.114460122 seconds
       [2017-05-02T18:48:57+00:00] FATAL: Stacktrace dumped to /tmp/kitchen/cache/chef-stacktrace.out
       Starting Chef Client, version 13.1.6
       resolving cookbooks for run list: ["chef_client_updater::default"]
       Synchronizing Cookbooks:
         - chef_client_updater (1.0.2)
         - compat_resource (12.19.0)
       Installing Cookbook Gems:
       Compiling Cookbooks...
       Converging 1 resources
       Recipe: chef_client_updater::default
         * chef_client_updater_updater[update chef-client] action update (up to date)

       Running handlers:
       Running handlers complete
       Chef Client finished, 0/1 resources updated in 08 seconds
       Finished converging <default-ubuntu-1404> (1m11.85s).
</details>

**Comment by tas50** [comment]
@mckownam What version of compat_resource cookbook are you running and does this occur if chef_client_updater is the only thing in your run list?

**Comment by mckownam** [comment]
compat_resource version = 12.19.0

Yes, this happens even if chef_client_updater is the only thing in the run list. I have also tested this in Ubuntu 14.04 and got the same error. 

The details of the setup are as follows:
 - Systems are all VMs running in VMware (and Chef is used for deploying patches, updates, software, etc)
 - Testing of this chef_client_updater cookbook came about as a result of the omnibus_updater and omnibus_updater_windows cookbooks being depreciated and potentially nonviable for Chef 13, so assessment of other methods of updating chef client began (and it was quickly determined that this cookbook did not work on Windows VMs, although that may be changed in the future; there was already another issue request about that when I last checked). 
 - The chef_client_updater cookbook is wrapped into a role in which the prevent_downgrade and version variables are set. 
 - Chef client 12.4.3 is the oldest version in our deployment (and 12.16.42 is the latest), so those two versions were picked to run the upgrade/downgrade tests against (upgrade tests from 12.16.42 went to 13.0.118). 
 - The chef nodes (ubuntu 12 and 14) started at 12.16.42 at the beginning of this test, then had their runlists replaced by the chef_client_updater role which pointed them to upgrade to 13.0.118. Upgrades were performed without issue. Then they were downgraded back to 12.16.42. No problems. 
 - Downgrades went from 12.16.42 to 12.4.3 and were successful. Upgrades from 12.4.3 to 12.16.42 resulted in the Mixlib error discussed in this thread and seen in the stacktrace.out below. 

Contents of stacktrace.out:
<pre>Generated at 2017-05-11 09:45:07 -0400
Mixlib::ShellOut::ShellCommandFailed: chef_client_updater_updater[update chef-client] (chef_client_updater::default line 20) had an error: Mixlib::ShellOut::ShellCommandFailed: chef_gem[mixlib-install] (/var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb line 45) had an error: Mixlib::ShellOut::ShellCommandFailed: Expected process to exit with [0], but received '1'
---- Begin output of /opt/chef/embedded/bin/gem install mixlib-install -q --no-rdoc --no-ri -v "3.2.1" ----
STDOUT: 
STDERR: ERROR:  While executing gem ... (TypeError)
    no implicit conversion of nil into String
---- End output of /opt/chef/embedded/bin/gem install mixlib-install -q --no-rdoc --no-ri -v "3.2.1" ----
Ran /opt/chef/embedded/bin/gem install mixlib-install -q --no-rdoc --no-ri -v "3.2.1" returned 1
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/mixlib-shellout-2.1.0/lib/mixlib/shellout.rb:289:in `invalid!'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/mixlib-shellout-2.1.0/lib/mixlib/shellout.rb:276:in `error!'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/mixin/shell_out.rb:56:in `shell_out!'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/provider/package.rb:501:in `shell_out_with_timeout!'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/provider/package/rubygems.rb:541:in `install_via_gem_command'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/provider/package/rubygems.rb:522:in `install_package'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/provider/package.rb:93:in `block in action_install'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/mixin/why_run.rb:52:in `call'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/mixin/why_run.rb:52:in `add_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/provider.rb:175:in `converge_by'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/provider/package.rb:92:in `action_install'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/provider.rb:140:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/resource.rb:584:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/resource/chef_gem.rb:60:in `block in after_created'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/resource/chef_gem.rb:59:in `each'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/resource/chef_gem.rb:59:in `after_created'
/var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/resource_builder.rb:106:in `build'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/dsl/recipe.rb:110:in `build_resource'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/dsl/recipe.rb:71:in `declare_resource'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/dsl/resources.rb:15:in `chef_gem'
/var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:45:in `rescue in load_mixlib_install'
/var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:42:in `load_mixlib_install'
/var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:64:in `mixlib_install'
/var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:33:in `block (2 levels) in class_from_file'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/mixin/why_run.rb:52:in `call'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/mixin/why_run.rb:52:in `add_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/provider.rb:175:in `converge_by'
/var/chef/cache/cookbooks/chef_client_updater/resources/updater.rb:32:in `block in class_from_file'
(eval):2:in `block in action_update'
/var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:132:in `instance_eval'
/var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/copied_from_chef/chef/provider.rb:132:in `compile_and_converge_action'
(eval):2:in `action_update'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/provider.rb:140:in `run_action'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/resource.rb:584:in `run_action'
/var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:78:in `run_action'
/var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `block (2 levels) in converge'
/var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `each'
/var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:106:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/resource_collection/resource_list.rb:83:in `block in execute_each_resource'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/resource_collection/stepable_iterator.rb:116:in `call'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/resource_collection/stepable_iterator.rb:116:in `call_iterator_block'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/resource_collection/stepable_iterator.rb:85:in `step'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/resource_collection/stepable_iterator.rb:104:in `iterate'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/resource_collection/stepable_iterator.rb:55:in `each_with_index'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/resource_collection/resource_list.rb:81:in `execute_each_resource'
/var/chef/cache/cookbooks/compat_resource/files/lib/chef_compat/monkeypatches/chef/runner.rb:105:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/client.rb:654:in `block in converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/client.rb:649:in `catch'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/client.rb:649:in `converge'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/client.rb:688:in `converge_and_save'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/client.rb:269:in `run'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/application.rb:270:in `block in fork_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/application.rb:258:in `fork'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/application.rb:258:in `fork_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/application.rb:224:in `block in run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/local_mode.rb:39:in `with_server_connectivity'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/application.rb:212:in `run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/application/client.rb:402:in `block in interval_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/application/client.rb:392:in `loop'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/application/client.rb:392:in `interval_run_chef_client'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/application/client.rb:382:in `run_application'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/lib/chef/application.rb:60:in `run'
/opt/chef/embedded/lib/ruby/gems/2.1.0/gems/chef-12.4.3/bin/chef-client:26:in `<top (required)>'
/usr/bin/chef-client:54:in `load'
/usr/bin/chef-client:54:in `<main>'</pre>

**Comment by lamont-granquist** [comment]
compat_resource is getting ripped out in #16 

this should probably get retested after #16 lands, because any bugs will likely be dramatically different bugs.

**Comment by wjordan** [comment]
I managed to track down the root cause of this issue, here is a minimal repro:

`Dockerfile`:
```Dockerfile
FROM ubuntu:16.04
ADD ./repro.sh /root
RUN /root/repro.sh
```

`repro.sh`:
```BASH
#!/bin/bash
apt-get update
apt-get install -y -q curl
install_chef() {
  curl -L https://omnitruck.chef.io/install.sh | bash -s -- -v $1
}
install_chef 12.16.42
/opt/chef/embedded/bin/gem install mixlib-install
install_chef 12.4.3
/opt/chef/embedded/bin/gem install mixlib-install
```

This issue relates to a bug (or rather, partial backwards-incompatibilities) in Rubygems binstub scripts.

### Summary
Rubygems uses a _very fragile_ string-matching regex to parse the spec name from an existing binstub:

- In versions `< 2.5.2` (rubygems/rubygems#1407), the regex Rubygems uses to match binstubs checks for `gem '[name]',` only.
- From version `>= 2.5.2` (rubygems/rubygems#1407) through `< 2.7.0` (rubygems/rubygems#1904), Rubygems generated binstubs without `gem '#{spec.name}',` contents.

Further compounding the problem, a bug (rubygems/rubygems#1560) caused `no implicit conversion of nil into String` error to be thrown when the binstub didn't match its regex. This bug affected versions < `2.6.3` (fixed in rubygems/rubygems#1561).

First, Chef `12.16.42` installs Rubygems `2.6.8`, which installs `mixlib-install` with a binstub missing the `gem 'mixlib-install',` magic-string. Next, Chef `12.4.3` installs Rubygems `2.4.8`, throwing the error on the next `mixlib-install` attempt.

### Solution

adding `options '--force'` to the `chef_gem` resource should fix this issue. It works around the bug by [bypassing the faulty regex-check logic entirely](https://github.com/rubygems/rubygems/blob/v2.4.8/lib/rubygems/installer.rb#L157).
